commit 4a6a8b056f39079d5e958eac84c2ad173f4680bc
Author: Philip Hazel <Philip.Hazel@gmail.com>
Date:   Fri Mar 1 17:23:34 2024 +0000

    Fix incorrect compiling when variable-length lookbehind's first branch was not the shortest and there was a backreference to a groupwithin the lookbehind.

diff --git a/ChangeLog b/ChangeLog
index 0568108a..6b78659e 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -4,6 +4,16 @@ Change Log for PCRE2
 Before the move to GitHub, this was the only record of changes to PCRE2. Now
 there is often more detail in the pull requests.
 
+Version 10.44 xx-xxx-2024
+-------------------------
+
+1. If a pattern contained a variable-length lookbehind in which the first 
+branch was not the one with the shortest minimym length, and the lookbehind 
+contained a capturing group, and elsewhere in the pattern there was another 
+lookbehind that referenced that group, the pattern was incorrectly compiled, 
+leading to unpredictable results, including crashes in JIT compiling. An 
+example pattern is: /(((?<=123?456456|ABC)))(?<=\2)/
+
 
 Version 10.43 16-February-2024
 ------------------------------
diff --git a/src/pcre2_compile.c b/src/pcre2_compile.c
index 95c97710..e66e40f6 100644
--- a/src/pcre2_compile.c
+++ b/src/pcre2_compile.c
@@ -9909,7 +9909,7 @@ do
   *bptr |= branchlength;  /* branchlength never more than 65535 */
   bptr = *pptrptr;
   }
-while (*bptr == META_ALT);
+while (META_CODE(*bptr) == META_ALT);
 
 /* If any branch is of variable length, the whole lookbehind is of variable
 length. If the maximum length of any branch exceeds the maximum for variable
diff --git a/testdata/testinput2 b/testdata/testinput2
index 97608542..9b4c0684 100644
--- a/testdata/testinput2
+++ b/testdata/testinput2
@@ -6099,4 +6099,8 @@ a)"xI
 /(a(?1)z||(?1)++)$/
     abcd\=disable_recurseloop_check
 
+/(((?<=123?456456|ABC)))(?<=\2)../
+    ABCDEFG
+    12345645678910 
+
 # End of testinput2
diff --git a/testdata/testoutput2 b/testdata/testoutput2
index ee1f70b0..30c5a83e 100644
--- a/testdata/testoutput2
+++ b/testdata/testoutput2
@@ -18038,6 +18038,16 @@ Failed: error -47: match limit exceeded
  0: 
  1: 
 
+/(((?<=123?456456|ABC)))(?<=\2)../
+    ABCDEFG
+ 0: DE
+ 1: 
+ 2: 
+    12345645678910 
+ 0: 78
+ 1: 
+ 2: 
+
 # End of testinput2
 Error -70: PCRE2_ERROR_BADDATA (unknown error number)
 Error -62: bad serialized data
